server/src/main/java/org/elasticsearch/search/fetch/PreloadedFieldLookupProvider.java

```diff
@@ -32,12 +32,25 @@ class PreloadedFieldLookupProvider implements LeafFieldLookupProvider {
 
     @Override
     public void populateFieldLookup(FieldLookup fieldLookup, int doc) throws IOException {
-        String field = fieldLookup.fieldType().name();
+        if (storedFields == null) {
+            loadDirect(fieldLookup, doc);
+            return;
+        }
+        final String field = fieldLookup.fieldType().name();
```

Comment:
could you help me understand what exactly triggers the bug you found? The design of the existing PreloadedFieldLookupProvider is not great. It makes some assumptions: storedFields is not null, setNextReader is called before populateFieldLookup. Which of these assumptions no longer hold with the changes made in #106325 ? I understand we are now retrieving stored fields from the FetchFieldsPhase, and things depend on ordering of subfetch phases, but I don't follow precisely the root cause.

`FetchPhaseDocsIterator` sets both the next reader as well as stored fields in nextDoc, and is shared across all the subfetch phases? What am I missing?

Reply:
I can't reproduce the issue locally...even if it failed a few times over tens of times I tried as soon as I try again it does not fail anymore...so my fix is based on the fact that sometimes `storedFields` is null which means caches are not loaded and it looks reasonable to me to fallback to direct loading in that case. Also considering that in our code I only see a single place where `storedFields` is set which is in `FetchPhase` when doing:

```
fieldLookupProvider.storedFields = hit.loadedFields();
```

Reply:
I don't feel good about this. I think we should dig deeper and understand where failures come from. The proposed fix may be hiding some other problem, as I believe we do expect storedFields to be set and not be null. I think it's important to understand why they may be null in some situations.